src/MyArrayList.java

```diff
@@ -0,0 +1,299 @@
+    public void add(Object elem) {
+        add(count, elem);
+    }
+
+    @Override
+    public void add(int index, Object elem) {
+        if (index < count) {
+            shiftRight(index);
+            array[index] = elem;
+        }
+
+        else if (index < array.length) {
+            array[index] = elem;
+        }
+
+        else {
+            expand(index);
+            array[index] = elem;
+        }
+
+        count = index > count ? ++index : ++count;
+    }
+
+    @Override
+    public Object set(int index, Object elem) {
+
+        Object temp;
+
+        if (index >= array.length) {
+              expand(index);
+              array[index] = elem;
+              temp = null;            //здесь мог быть Exception
+          }
+
+         else {
+              temp = array[index];
+              array[index] = elem;
+          }
+
+        count = index > count ? ++index : count;
+
+        return temp;
+    }
+
+    @Override
+    public boolean remove(Object elem) {
+        int temp = indexOf(elem);
+
+        if (temp == -1) {
+            return false;
+        }
+
+        else {
+            remove(temp);
+            return true;
+        }
+
+    }
+
+    @Override
+    public Object remove(int index) {
+        Object temp = this.array[index];        //здесь может быть Exception
+        shiftLeft(index);
+        count--;
+
+        return temp;
+    }
+
+    @Override
+    public void clear() {
+        Object[] array = new Object[this.array.length];
+        Arrays.fill(array, 0, this.array.length, null);
+        this.array = array;
+        count = 0;
+    }
+
+    @Override
+    public Object get(int index) {
+        return array[index];            //здесь может быть Exception
+    }
+
+    @Override
+    public int indexOf(Object elem) {
+        for (int i = 0; i < array.length; i++) {
+            if (elem == null) {
+                if (array[i] == null) {
+                    return i;
+                }
+            }
+
+         else {
+              if (elem.equals(array[i])) {
+                  return i;
+                }
+            }
+
+        }
+
+        return -1;
+    }
+
+    @Override
+    public boolean contains(Object elem) {
+        return indexOf(elem) != -1;
+    }
+
+    @Override
+    public int size() {
+        return count;
+    }
+
+    @Override
+    public Iterator iterator() {
+        return new MyIterator();
+    }
+
+
+    private class MyIterator implements Iterator {
+
+        private int iteratorCount = 0;
+
+        @Override
+        public boolean hasNext() {
+            return count > iteratorCount;
+        }
+
+        @Override
+        public Object next() {
+            return array[iteratorCount++];          //здесь может быть Exception
+        }
+
+        @Override
+        public void remove() {
+            MyArrayList.this.remove(--iteratorCount);           //здесь может быть ArrayIndexOutOfBoundsException
+            iteratorCount++;
+        }
```

Comment:
For remove there is a must-have condition: we have to call next() before calling remove().

This code should work: 
`while(itr.hasNext()) {
    itr.next();
    itr.remove();
}`

This code should NOT work: 
`while(itr.hasNext()) {
    itr.remove(); // should throw any RuntimeException e.g. IllegalStateException
}`